crates/turborepo-lib/src/graph/walker.rs

```diff
@@ -0,0 +1,257 @@
+use std::{collections::HashMap, hash::Hash};
+
+use futures::future::join_all;
+use petgraph::{
+    visit::{IntoNeighborsDirected, IntoNodeIdentifiers},
+    Direction,
+};
+use tokio::{
+    sync::{broadcast, mpsc, oneshot, watch},
+    task::JoinHandle,
+};
+use tracing::log::trace;
+
+pub struct Walker<N> {
+    cancel: watch::Sender<bool>,
+    node_events: Option<mpsc::Receiver<(N, oneshot::Sender<()>)>>,
+    join_handles: Vec<JoinHandle<()>>,
+}
+
+pub type WalkMessage<N> = (N, oneshot::Sender<()>);
+
+// These constraint might look very stiff, but since all of the petgraph graph
+// types use integers as node ids and GraphBase already constraints these types
+// to Copy + Eq so adding Hash + Send + 'static as bounds aren't unreasonable.
+impl<N: Eq + Hash + Copy + Send + 'static> Walker<N> {
+    /// Create a new graph walker for a DAG that emits nodes at a given
+    /// The graph should not be modified after a walker is created as the nodes
```

Comment:
Is there a way to ensure that we have an exclusive reference to the graph?

Reply:
I think lifetimes could get hairy quite quickly but maybe we should have the Walker hold a reference to the graph for its entire lifetime to ensure that it cannot be mutably borrowed elsewhere.

Reply:
I'm slightly opposed to adding this restriction here as is clashes with the approach that `petgraph` takes for graph traversals
> [Dfs](https://docs.rs/petgraph/latest/petgraph/visit/struct.Dfs.html), [Bfs](https://docs.rs/petgraph/latest/petgraph/visit/struct.Bfs.html), [DfsPostOrder](https://docs.rs/petgraph/latest/petgraph/visit/struct.DfsPostOrder.html) and [Topo](https://docs.rs/petgraph/latest/petgraph/visit/struct.Topo.html) are basic visitors and they use “walker” methods: the visitors don’t hold the graph as borrowed during traversal, only for the .next() call on the walker. They can be converted to iterators through the [Walker](https://docs.rs/petgraph/latest/petgraph/visit/trait.Walker.html) trait.

[source](https://docs.rs/petgraph/latest/petgraph/visit/index.html#graph-traversal)

Walking doesn't actually need to access the underlying graph once the struct has been built, it just will send node ids out in the order. It is up to the caller to use those node ids how they wish. If people feel strongly that we should disallow this, then I can look into holding some reference to the graph.

Reply:
I think we can wait on this until we have more concrete use-cases and see if we want the restriction